File: leetcode/257.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> pathToLeaves = new ArrayList<>();
        traverse(root, new StringBuilder(), pathToLeaves);
        return pathToLeaves;
    }

    private void traverse(TreeNode node, StringBuilder sb, List<String> path) {
        if (node == null) {
            return;
        }
        if (node.left == null && node.right == null) {
            path.add(sb.append(node.val).toString());
            return;        
        }
        int length = sb.length();
        traverse(node.left, sb.append(node.val).append("->"), path);
        sb.setLength(length);
        traverse(node.right, sb.append(node.val).append("->"), path);
        sb.setLength(length);
    }
}

File: code/257.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        pathToLeaves = []
        
        def traverse(node, paths):
            if node == None:
                paths.append(None)
                return
            paths.append(node.val)
            if (node.left == None and node.right == None):
                pathToLeaves.append(paths.copy())
                return
            traverse(node.left, paths)
            paths.pop()
            traverse(node.right, paths)
            paths.pop()

        traverse(root, [])
        all_paths = []
        for p in pathToLeaves:
            p = map(str, p)
            all_paths.append("->".join(p))
        return all_paths
        